### DISCUSSION TITLE
Structured Citations from Native Functions

### DISCUSSION THREAD

**Discussion by htaljaard** [microsoft/semantic-kernel#8884] [Q&A] [answered]
Hey gang,

I am looking for a way to produce structured citations in my responses from native functions (Like AI Search does for the Azure OpenAI SDK).

I have tried a few things (including trying to structure the prompt output) but the results are inconsistent. 
I also tried adding an empty list of citations in the main function and adding List<Citation> as a parameter to the plugin. 

Then I found the ```Data``` property of the Kernel.

This gives me what I need. 

```
public async Task<List<Citation>> GetUnitGoals(string unitName, string indexName,
[Description("Examples include Unit Goals, Content Descriptions, Inquiry Skills")] string contentType,

[Description("A list of citations that needs to be updated with the search results.")] List<Citation> citations,
Kernel kernel)
{
    var results = await _azureAISearchService.SearchAsync(query: $"{unitName} {contentType}", indexName: indexName, vectorQuery: null);

    var localCitations = results.Select(r => new Citation(Content: r.Content, Id: r.Id, Title: r.FileName, Filepath: r.Url, Url: r.Url, ChunkId: r.ChunckId.ToString())).ToList();

    kernel.Data.Add(KernelDataConstants.Citations, localCitations);

    return localCitations;
}

```

This example uses AI Search, but the same will requirement apply to Bing and whatever we come up with in the future. Bing is what prompted this requirement as it is retrieval, but not "memory" per say. 

Is it safe to use the ```Data``` Property or is there a better way?

Thanks in advance

**Comment by markwallace-microsoft** [comment, marked as answer]
Hi @htaljaard 
you can use the `Kernel.Data` property once you are creating transient instances of the `Kernel` so you avoid have citations from one request overwriting those from another.
Is your `GetUnitGoals` function being invoked via LLM function calling? If it is you have the option to terminate the function calling loop when this function gets called and than return this function response as the final response from the LLM.

**Reply by htaljaard** [reply]
HI @markwallace-microsoft 

Thanks! that is great news. 

The get unit Goals typically forms part of a larger prompt (build lesson plan) - so I love the idea of terminating the loop after a specific function - that would be exactly what I need for the lesson plan prompt.

The lesson plan prompt is not a native function though
![image](https://github.com/user-attachments/assets/1e6655db-0ab9-42d6-9651-492def1df672)

I tried looking online quickly, but don't see how to terminate the loop. Can you please advise?

**Reply by htaljaard** [reply]
ok, after a bit of additional research, I saw Filters. I assume we need to use those to stop the flow. Thanks @markwallace-microsoft

**Comment by rutzsco** [comment]
I need to do something similar in Python, but I don't see Kernel.Data. Is there an alternative way to accomplish this in Python? Thanks!
